add validation

apomarinov opened the pull request:
- Request Validation - refer to [rpg](https://github.com/prod9/rabbit-pg-adapter)
  - `httpin` used to read request data into structs
  - `govalidator` used for validation
  - validate request body and query params
  - custom validation messages for `govalidator`; TODO: add custom validations
  - conditional validation based on header `x-api-action`; create/update/etc
  - add err directly to `render.JSON` to skip if/else in ctrl

apomarinov committed:
contrib: Add validation

Poowis-Prodigy9 commented on httpserver/middlewares/validator.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// errorHandler will unwrap httpin error to pass the fx validation errors to render
+func errorHandler(rw http.ResponseWriter, r *http.Request, err error) {
```
I ran into a problem with this function.
when I make a request with something like this
```
{
   "name":"Finance",
   "code": "FIN" 
   // comment
}
```
It doesn't render error from requesToStruct

chakrit commented:
I've finally got time to try out the changes. I think this PR's a bit un-idiomatic for Go:

1. This uses a lot of reflection
2. Prefers magic code (middleware does the validation and load into context) to drive business logic
3. `httpin` feels much too heavy for the use case we want and also not very minimialistic/simplistic in design.
4. There're a lot of code doing translations, with the current pattern we're doing in the comp, I think it's better done at the frontend using some node.js tech rather than adding such support on the api side of things.

I've repurposed the fieldmap idea in validation, got rid of deps and landed https://github.com/prod9/fx/blob/main/validate/validate.go

Not ideal yet, but it's a good start, with the following assumptions:

1. Modifications are the ones requiring validations the most so it's meant to be used in UOW Create/Update/Delete structs
2. Helper methods have been added to help w the boilerplate in controllers.
3. I've implemented a whole Todo app from scratch using the added code as example

* Using validate: https://github.com/prod9/fx/blob/main/examples/todoapi/auth/create_user.go#L18
* Invoking the UOW action in controller: https://github.com/prod9/fx/blob/main/examples/todoapi/auth/user_ctr.go#L26 (validation are automatically applied and error returned)

---

This PR is still useful though, I'm thinking maybe we can put code that are already used in the wild but not merged to main in a "contrib" sort of folder or repo, not sure exactly how yet.

chakrit committed:
contrib: Revert changes to base fx

chakrit committed:
fx: Adds an AddField helper to validate.Error

Since NewFieldError actually returns error (for boxing reasons) and so
it can't be directly used in validate.Error{}.Add without type assertion
which is weird to have to do from a user's perspective.

chakrit committed:
contrib: Retrofit apo's impl with base validation error package
